feature(edquot): Add quota configuration and EndOfQuotaNemesis

Mark-Gurevich opened the pull request:
According to task: https://github.com/scylladb/qa-tasks/issues/1086

This commit adds the possibility to configure quota on a node and adds the EndOfQuotaNemesis
The EndOfQuotaNemesis configures quota with current disk usage + 1GB and waits for `Disk quota exceeded` message in logs. Then the quota is cleared and scylla-server service is restarted.

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

fruch commented on sdcm/nemesis.py:
```diff
@@ -1661,6 +1662,52 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+                    configure_quota_on_node(target_node=node)
+                    wait.wait_for_log_lines(node=node, start_line_patterns='Disk quota exceeded',
+                                            end_line_patterns='Will retry in 3200ms',
+                                            start_timeout=60, end_timeout=600)
```
The time 10min and size 1gb, are highly dependent on the test case, in some it would take more then 10min to fill 1gb.

Anyhow it's not clear what the end message you are looking for, it's a bit vague, and not clear why it's needed.

Mark-Gurevich replied:
Anyway, the `wait_for_log_lines` didn't work at all. 
In the last run it jumped right away to the next method `clean_edquota_on_node` so something it wrong here.

fruch replied:
I believe it's a context manger, so you are using it the wrong way

Mark-Gurevich replied:
Yep, I saw that already. Thank you for the input.

fruch commented on sdcm/nemesis.py:
```diff
@@ -6409,3 +6456,11 @@ class DisableBinaryGossipExecuteMajorCompaction(Nemesis):
         self.disrupt_disable_binary_gossip_execute_major_compaction()
+
+
+class EdquotNemesis(Nemesis):
```
Name of the class should be similar to the name of the distrupt function
Edquot is clear or known enough, from my POV.

The function name is clearer

Mark-Gurevich replied:
I haven't fully got you. Should I change the `disrupt_end_of_quota_nemesis` to `disrupt_edquot_nemesis` you mean ?

fruch replied:
I think the other way around `EndOfQoutaNemesis` has more meaning to me then `EdquotNemesis` sounds like a typo at first look.

yeah, I know we have `EnospcNemesis`, but that name was like that before I joined scylla...
`NoSpaceNemesis` would have been more readable (especially for newcomers)

Mark-Gurevich replied:
I totally agree with you. 
I've used the `EDQUOT` because that's how it is described [here](https://github.com/scylladb/qa-tasks/issues/1086) and [here](https://github.com/scylladb/scylladb/pull/12653)
I guess its a known term in Linux.
Whatever you decide, no problem to change this tiny thing.

Mark-Gurevich commented:
Do I need to set the new `enable_quota` to `false` anywhere in the `defaults` for example in `test_default.yaml` ?

fruch commented:
> Do I need to set the new `enable_quota` to `false` anywhere in the `defaults` for example in `test_default.yaml` ?

yes, it's best if any newly introduced parameter has a default in test_default.yaml, or in the different backend defaults files.
there should be any defaults set in the code, default should be in once place (or per backend)

Mark-Gurevich commented:
> > Do I need to set the new `enable_quota` to `false` anywhere in the `defaults` for example in `test_default.yaml` ?
> 
> yes, it's best if any newly introduced parameter has a default in test_default.yaml, or in the different backend defaults files. there should be any defaults set in the code, default should be in once place (or per backend)

As I expected. Added to test_default.yaml `enable_quota: false`

soyacz commented on sdcm/nemesis.py:
```diff
@@ -1661,6 +1662,63 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+
+        for node in nodes:
+            if self._is_it_on_kubernetes():
+                LOGGER.info("Skipping nemesis for kubernetes")
```
raise UnsupportedNemesis

Mark-Gurevich replied:
ok

soyacz commented on sdcm/nemesis.py:
```diff
@@ -1661,6 +1662,63 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+            with ignore_disk_quota_exceeded_errors(node=node):
+                result = self.target_node.remoter.run('cat /proc/mounts')
+                if '/var/lib/scylla' not in result.stdout:
+                    LOGGER.error("Scylla doesn't use an individual storage, skip edquota test")
```
raise UnsupportedNemesis

Mark-Gurevich replied:
Changed. Why we don't raise in enospc nemesis in the same place then ?

soyacz commented on sdcm/sct_config.py:
```diff
@@ -1484,6 +1484,9 @@ class SCTConfiguration(dict):
              help="""Use dns names instead of ip addresses for nodes in cluster"""),
 
+        dict(name="enable_quota", env="SCT_ENABLE_QUOTA", type=bool,
+             help="Enable possibility of quota configuration on the cluster during setup")
```
why it is required to enable like this?

Mark-Gurevich replied:
Actually we can enable quota on node during the nemesis and restart the node.
@fgelcer asked me to make it this way, so when it is enabled on the whole cluster we can use the nemesis or `configure_quota_on_node_for_scylla_user` separately.

roydahan replied:
All this is not needed.

soyacz commented on sdcm/nemesis.py:
```diff
@@ -1661,6 +1662,63 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+                            throw_exc=False,
+                        ), "Quota has not been reached"
+                    else:
+                        LOGGER.error("Failed to configure quota on node: {}".format(node))
```
raise UnsupportedNemesis or other error if appropriate

soyacz commented on sdcm/utils/common.py:
```diff
@@ -2756,6 +2756,54 @@ def clean_enospc_on_node(target_node, sleep_time):
+            return False
+
+
+def clean_edquota_on_node(target_node, sleep_time):
```
why do we need `sleep_time`?

soyacz commented on sdcm/nemesis.py:
```diff
@@ -1661,6 +1662,63 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+            if self._is_it_on_kubernetes():
+                LOGGER.info("Skipping nemesis for kubernetes")
+                break
+            node.verify_quota_is_enabled_on_node()
```
`verify_quota_is_configured_for_scylla_user` and `configure_quota_on_node` use target node as argument, why here is different?

Mark-Gurevich replied:
Why we have methods in `BaseScyllaCluster` and `BaseNode` classes and other methods are in sdcm/utils/common.py? 
`verify_quota_is_enabled_on_node` is related to node setup and verify the possibility of adding quota configuration (enabled). So, I added it to `BaseNode` class.
Maybe I should change the name for `configure_quota_on_node` to `configure_quota_on_node_for_scylla_user`.
Anyway, both these methods (`verify_quota_is_configured_for_scylla_user` and `configure_quota_on_node`) are in the utils/common.py and require a parameter as target_node.
That's the logic of SCT which is not very clear to me why there is a division between `BaseNode` and methods in `common.py` and other files in `sdcm/utils`.

soyacz commented on sdcm/nemesis.py:
```diff
@@ -1661,6 +1662,63 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+                        quota_exceeded = node.follow_system_log(patterns=['Disk quota exceeded'])
+                        LOGGER.debug('Run additional load with prepare_test_table to reach end of quota. '
+                                     'Create keyspace1.standard1 if it does not exist')
+                        self._prepare_test_table(ks='keyspace1', table='standard1')
```
this will not fill disk anyway if table already exists

soyacz commented on sdcm/utils/common.py:
```diff
@@ -2756,6 +2756,54 @@ def clean_enospc_on_node(target_node, sleep_time):
     target_node.wait_db_up()
 
 
+def verify_quota_is_configured_for_scylla_user(target_node) -> bool:
```
put it to separate file rather than `common.py` - it is not that common (maybe smthing like `xfs_quota.py`?)

Mark-Gurevich replied:
I can put it to separate file but why we don't create a separate file for enospc methods then?

soyacz commented:
I wonder if creating 1GB file with dd and changing ownership to `scylla` would do the job instead creating kesypace, table and populating it...

fgelcer commented:
> I wonder if creating 1GB file with dd and changing ownership to `scylla` would do the job instead creating kesypace, table and populating it...

well, it would, but as we should not login as Scylla user, we won't use the right user quota, and won't test the behavior on Scylla's side

soyacz commented:
> > I wonder if creating 1GB file with dd and changing ownership to `scylla` would do the job instead creating kesypace, table and populating it...
> 
> well, it would, but as we should not login as Scylla user, we won't use the right user quota, and won't test the behavior on Scylla's side

we don't have to login as Scylla, just change file ownership.

roydahan commented on sdcm/nemesis.py:
```diff
@@ -1665,6 +1666,77 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+    def write_data_with_cs_to_reach_end_of_quota(self, node):
+        keyspace_quota = 'ks_quota'
+        try:
+            stress_cmd = "cassandra-stress write no-warmup n=200000 cl=QUORUM -mode native cql3 " \
+                         f"-schema 'replication(strategy=NetworkTopologyStrategy," \
+                         f"replication_factor={self.tester.reliable_replication_factor})' -log interval=5"
+            self.tester.run_stress_thread(
+                stress_cmd=stress_cmd, keyspace_name=keyspace_quota, stop_test_on_failure=False, round_robin=True)
```
.

Mark-Gurevich commented:
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-5gb-1h-EndOfQuotaNemesis-aws/15/

Mark-Gurevich commented:
> https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-5gb-1h-EndOfQuotaNemesis-aws/15/

The test is working. I checked the logs and verified it has actually configured quota, got the expected message, went down and restarted Scylla service.

roydahan commented on sdcm/nemesis.py:
```diff
@@ -1688,6 +1691,76 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
                     finally:
                         clean_enospc_on_node(target_node=node, sleep_time=sleep_time)
 
+    def write_data_with_cs_to_reach_end_of_quota(self, node):
```
I already commented long ago that this is not needed and what is need to be done is to write a sparse file as we do in ENOSPC nemesis.

roydahan replied:
Not only it's not needed, it adds a load that some of the tests may not expect.

roydahan replied:
Still relevant.

roydahan commented on sdcm/nemesis.py:
```diff
@@ -1665,6 +1666,77 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+        except IOError as exc:
+            LOGGER.info("We should have reached the expected I/O error and quota has exceeded\n Message: {}".format(str(exc)))
+        finally:
+            LOGGER.debug("Dropping keyspace: %s", keyspace_quota)
+            with self.cluster.cql_connection_patient(node=self.target_node) as session:
+                session.execute(f""" DROP KEYSPACE IF EXISTS {keyspace_quota} """)
```
related to my previous comment.... while this is a month old comment.

roydahan commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,70 @@
+    node.remoter.sudo(cmd=enable_uquota_on_mount_cmd)
+    LOGGER.debug('Rebooting node: "{}"'.format(node.name))
+    node.reboot(hard=False)
+    LOGGER.debug("node reboot takes very long time and next methods fails. "
+                 "Adding sleep for 2 minutes here before wait_node_fully_start")
+    time.sleep(120)
```
Shouldn't be like that, at the maximum you should do "wait_for" and increase timeout.

roydahan commented on unit_tests/test_nemesis_sisyphus.py:
```diff
@@ -71,7 +71,7 @@ def test_list_all_available_nemesis(generate_file=True):
         subclasses_list=subclasses, export_properties=True)
 
-    assert len(disruption_list) == 84
+    assert len(disruption_list) == 85
```
Missing output of nemesis.yaml with the new nemesis.

roydahan commented on sdcm/nemesis.py:
```diff
@@ -1688,6 +1691,76 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+            self.write_data_with_cs_to_reach_end_of_quota(node)
+
+            remove_quota_configuration_from_node(target_node=node)
+            LOGGER.debug('Sleep a while before restart scylla-server')
+            time.sleep(sleep_time / 2)
```
Ugly and I don't understand why needed.

roydahan commented on data_dir/nemesis.yml:
```diff
@@ -332,3 +332,7 @@
+- disrupt_end_of_quota_nemesis:
+  - disruptive = True
+  - kubernetes = False
+  - free_tier_set = True
```
I think this one isn't relevant for this nemesis.

roydahan commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,76 @@
+LOGGER = logging.getLogger(__name__)
+
+
+def configure_quota_on_node_for_scylla_user(target_node):
```
He wasn't able to make it work for directory quota, so the workaround was to use a user quota.

Mark-Gurevich commented:
I hope this can be finally merged.
https://argus.scylladb.com/workspace?state=WyI4NWJiY2UxZC1iYmVhLTQ3ZGItOThkNS01NzExOWM3ZjQyNWIiXQ
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-5gb-1h-EndOfQuotaNemesis-aws/23/

soyacz commented on sdcm/nemesis.py:
```diff
@@ -1688,6 +1692,42 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+
+        with ignore_disk_quota_exceeded_errors(node=node):
+            quota_size = configure_quota_on_node_for_scylla_user(node=node)
+            verify_quota_is_configured_for_scylla_user(node=node)
```
if this fails, then quota on node will persist.
One proposal is to use try..catch or better to make `configure_quota_on_node_for_scylla_user` to be context manager which removes quota upon `__exit__`

soyacz commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,113 @@
+    LOGGER.info("Quota was configured for scylla user with value: {}".format(quota_value))
+
+
+def remove_quota_configuration_from_node(node):
```
this doesn't remove quota configuration from node, rather from user. Fix confusing name.

soyacz commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,113 @@
+        node.wait_node_fully_start()
+
+
+def verify_quota_is_enabled_on_node(node):
```
`verify_` in the name points to verification -> should not return anything (like it doesn't `verify_quota_is_configured_for_scylla_user`)
To keep consistency in naming, use `is_quota_enabled_on_node` for returning quota status

soyacz commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,113 @@
+        occupy_space_size = int((quota_size - currently_used_space) * 90 / 100)
+        occupy_space_cmd = f"su - scylla -c 'fallocate -l {occupy_space_size}K /var/lib/scylla/occupy_90percent.{time.time()}'"
+        with EventsSeverityChangerFilter(new_severity=Severity.NORMAL,  # End of quota created CRITICAL event
+                                         event_class=CassandraStressEvent,
```
how about other load tools like s-b?
Also, shouln't be done as a part of the `ignore_disk_quota_exceeded_errors` so in case c-s fails in between we don't create critical?

fruch replied:
why would c-s gonna fail ? or have critical events in this case ? i.e. its one node going down.

c-s need to continue without any issues, if not, we should open a bug, I don't think we do this during regular ENSPC nemesis

soyacz commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,113 @@
+        patterns=[r'Shutting down communications due to I\/O errors until operator intervention: Disk error:.*Disk quota exceeded'])
+    node.remoter.sudo('chsh -s /bin/bash scylla')  # Enable login for scylla user
+
+    def approach_end_of_quota():
```
why not just creating file that occupies 100%?

soyacz reviewed (commented):
Still some places to fix, but overall, looks good.
Generally good structure and namings, makes it easy to follow and review!

fruch commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,128 @@
+        currently_used_space = get_currently_used_space_by_scylla_user(node)
+        occupy_space_size = int((quota_size - currently_used_space) * 90 / 100)
+        occupy_space_cmd = f"su - scylla -c 'fallocate -l {occupy_space_size}K /var/lib/scylla/occupy_90percent.{time.time()}'"
+        # with EventsSeverityChangerFilter(new_severity=Severity.NORMAL,  # End of quota created CRITICAL event
```
can we remove this comment ?

soyacz commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,124 @@
+from contextlib import contextmanager
+
+from sdcm.exceptions import QuotaConfigurationFailure
+# from sdcm.sct_events import Severity
+# from sdcm.sct_events.filters import EventsSeverityChangerFilter
+# from sdcm.sct_events.loaders import CassandraStressEvent
```
remove these comments

soyacz reviewed (dismissed):
LGTM

Mark-Gurevich commented:
There is still a problem. If we wait for db to go down after  'Quota has exceeded' then, the test becomes flaky. It may fail as there is no space to flush memtables.
`[shard 1:main] table - Memtable flush failed due to: storage_io_error (Storage I/O error: 122: Disk quota exceeded). Will retry in 10000ms`
I've opened an [issue](https://github.com/scylladb/scylladb/issues/15908) on this, but Yaniv says it's expected behaviour. So the only option is to clean the sparse files immediately after we reach end of quota.
@roydahan Please advice.

![image](https://github.com/scylladb/scylla-cluster-tests/assets/13746611/32420980-41a6-4008-8fe3-a709cc01faa2)

Maybe we can try to ignore this part:
```
2023-10-31T12:39:15.188+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3      !ERR | scylla[582]:  [shard 1:main] table - failed to write sstable /var/lib/scylla/data/keyspace1/standard1-86364b8077e711eebea6fe9ae48a3174/me-3gao_0z5d_4ke7k2nn0ue3opyjr7-big-Data.db: storage_io>
2023-10-31T12:39:15.188+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3  !WARNING | scylla[582]:  [shard 1:main] table - Memtable flush failed due to: storage_io_error (Storage I/O error: 122: Disk quota exceeded). Will retry in 10000ms
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3  !WARNING | systemd[1]: scylla-server.service: State 'stop-sigterm' timed out. Killing.
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3   !NOTICE | systemd[1]: scylla-server.service: Killing process 582 (scylla) with signal SIGABRT.
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]: Aborting on shard 0.
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]: Backtrace:
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5a9efd8
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5ad5212
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   /opt/scylladb/libreloc/libc.so.6+0x3dbaf
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   /opt/scylladb/libreloc/libc.so.6+0x10ab5c
2023-10-31T12:39:22.938+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5afea40
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5afa7b2
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5afaedc
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5ab1eac
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5ab0f29
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5a51ff7
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x5a511ac
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x130e16a
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x130fbc0
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x130c6fc
2023-10-31T12:39:22.939+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   /opt/scylladb/libreloc/libc.so.6+0x27b89
2023-10-31T12:39:22.940+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   /opt/scylladb/libreloc/libc.so.6+0x27c4a
2023-10-31T12:39:22.940+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | scylla[582]:   0x130a124
2023-10-31T12:39:22.940+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | systemd[1]: Created slice Slice /system/systemd-coredump.
2023-10-31T12:39:22.940+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3     !INFO | systemd[1]: Started Process Core Dump (PID 1915/UID 0).
2023-10-31T12:39:22.940+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3  !WARNING | kernel: xfs filesystem being remounted at /run/systemd/unit-root/var/lib/systemd/coredump supports timestamps until 2038 (0x7fffffff)
2023-10-31T12:39:22.940+00:00 longevity-5gb-1h-EndOfQuotaNemesis--db-node-e21f0c0a-3  !WARNING | kernel: xfs filesystem being remounted at /run/systemd/unit-root/var/lib/systemd/coredump supports timestamps until 2038 (0x7fffffff)
```

It creates 3 Errors:
1. (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=666f8c6e-5ead-4ef4-9e36-598ac1b2858d: type=ABORTING_ON_SHARD regex=Aborting on shard 
2.  (CoreDumpEvent Severity.ERROR)
3.  (ClusterHealthValidatorEvent Severity.CRITICAL)

**I suggest just to remove the sparse files immediately after end of quota reached and only then wait for db to go down.**


Even after trying to remove the sparse files the service still fails to go down, and nemesis fails. 

![image](https://github.com/scylladb/scylla-cluster-tests/assets/13746611/6d9e2d49-b06c-4ccc-ba57-0eb67bdd86f8)

Mark-Gurevich commented:
Ready for merge
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-5gb-1h-EndOfQuotaNemesis-aws/30/

roydahan commented:
Please see my comment here: https://github.com/scylladb/scylladb/issues/15908#issuecomment-1793862000

fruch commented on sdcm/utils/quota.py:
```diff
@@ -0,0 +1,120 @@
+             "string to appear in the node log",
+             throw_exc=False
+             )
+    node.remoter.sudo('rm -rf /var/lib/scylla/occupy_90percent.*')  # Remove the sparse files
```
where is the code that clears this file ?

fruch replied:
or in other word, we need to make sure we always execute this part, this should be in one finally clause of some context manager, to make sure we clear those

fruch commented on sdcm/nemesis.py:
```diff
@@ -1688,6 +1691,40 @@ def disrupt_nodetool_enospc(self, sleep_time=30, all_nodes=False):
+        with ignore_disk_quota_exceeded_errors(node):
+            with configure_quota_on_node_for_scylla_user_context(node) as quota_size:
+                write_data_to_reach_end_of_quota(node, quota_size)
+            node.restart_scylla_server()
```
this is happening too quickly, you should align to the flow done in `clean_enospc_on_node`
it's first give something like 30sec for the scylla to shut down on it's own,
clears the temporary file
wait more 15sec
and then restarts scylla

if you are doing stop/restart on you own before scylla shut itself down.

you are hitting the systemd timeout, and core dumps scylla.

please stick to the same exact flow of ENOSPC, scylla behavior needs to be identical (from scylla code POV)
you trying total different flow here, is kind of making all kind of assumptions which aren't exactly accurate

roydahan replied:
Even longer than 30sec, we should make sure cluster react to the node failure.

Mark-Gurevich commented:
This is ready to be merged - passed in all variation with and without wait.
@scylladb/qa-maintainers

roydahan commented:
1. There are still some comments that aren't resolved.

2. We should merge it with enough sleep time during the problem (5 minutes), till the node is down.
IIUC, there is an issue with this - https://github.com/scylladb/scylladb/issues/15908
So if we merge this it may cause issues.

3. Let's send a followup PR to ENOSPC nemesis to also wait 5 minutes during the ENOSPC.

Mark-Gurevich commented:
> 1. There are still some comments that aren't resolved.

Which comments do you mean?
> 2. We should merge it with enough sleep time during the problem (5 minutes), till the node is down.
>    IIUC, there is an issue with this - [scylla server service shut down can't complete with `disrupt_end_of_quota_nemesis` running - due to lack of disk space to flush memtables scylladb#15908](https://github.com/scylladb/scylladb/issues/15908)
>    So if we merge this it may cause issues.

I've run tests with 5 and 10 minutes wait (runs 33,34) and they passed.

Israel wrote:
>if you are doing stop/restart on you own before scylla shut itself down.
>you are hitting the systemd timeout, and core dumps scylla.

The reason for the failure was probably trying to `stop_scylla_server` manually after it was already stopping when reached end of quota. 

> 3. Let's send a followup PR to ENOSPC nemesis to also wait 5 minutes during the ENOSPC.

Do you want me to change in the ENOSPC just the wait time to exact 5 minutes( why 5 ?)
Isn't it better to put some condition instead of relying on specific time?
Anyway, if we send a PR for ENOSPC with 5 minutes I should probably change to 5 minutes in this PR as well.
Right now it is 60 seconds.
@roydahan

fruch reviewed (dismissed):
LGTM

Waiting for handling those two:
- https://github.com/scylladb/scylla-cluster-tests/pull/6616#discussion_r1413484132
- stable 5min to let scylla go down, as @roydahan requested

Mark-Gurevich commented:
> LGTM
> 
> Waiting for handling those two:
> 
> * [feature(edquot): Add quota configuration and EndOfQuotaNemesis #6616 (comment)](https://github.com/scylladb/scylla-cluster-tests/pull/6616#discussion_r1413484132)

Added try - finally block
> * stable 5min to let scylla go down, as @roydahan requested

Changed to 5 mintues wait.

@fruch  - Test passed: 
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-5gb-1h-EndOfQuotaNemesis-aws/

Please merge.

soyacz reviewed (approved):
lgtm
